KEP: deployment controller to support in-place upgrade

jian-he opened the pull request:
A proposal for deployment controller to support in-place upgrade

adohe commented:
/cc @adohe

janetkuo commented:
/ok-to-test
/assign @janetkuo

xiang90 commented:
@kow3ns can you take a look? thanks!

kargakis commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+    // that oldReplicaSet may kill the pods prematurely.
+    pause(oldReplicaSet)
+    
+    // Reduce oldReplicaSet by the BatchSize
```
What is BatchSize?

jian-he replied:
// The number of pods to be updated in each batch.
This is mentioned in the API changes section.

kargakis commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+    for pod range(podsToUpdate)
+        update(pod, newReplicaSet-controllerRef)
+
+    // Increase newReplicaSet by BatchSize
```
Unfortunately this will not work.

1 - replica sets do not support pausing
2 - pausing (in the context of a deployment) means pausing an update. scale up/down still works as expected so if a new pod is added out of bound, it will get deleted automatically because now the deployment has more pods than expected.

jian-he replied:
Correct, if we go with approach 1, then we need to make replica set support it. So far, I feel approach 2 is better to go, as it's simpler

kargakis commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+
+    // swap oldReplicaSet and newReplicaSet information such as spec, revision, annotation etc.
```
This swap is not atomic and will not play well with the deployment controller and the mechanism it uses for creating new replicasets, though I know @janetkuo has a proposal to change that.

jian-he replied:
Could you elaborate what the issue is ? Since the new ReplicaSet is created with 0 replicas, IIUC, this behaves like a no-op, even though the swap is not atomic. the flow is like below:
1. oldCopy-info = oldReplicaSet-info // write the copy into etcd,  in case the controller failed, 
2. set(oldReplicaSet, newReplicaSet-info)
3. set(newReplicaSet, oldCopy-info)

kargakis commented:
@mfojtik @tnozicka @smarterclayton

xiang90 commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+    
+    // Pause the old replica set. This is required for this scenario: if we reduce oldReplicaSet immediately, there's a chance
```
replicaset does not support pause today. we could potentially add a pause functionality to pause the control of the controller. (based on my experience with building quite a few operators, pausing functionality is important for controllers that will be used/controlled by another higher level controller.)

kargakis replied:
sounds like a different proposal?

kargakis replied:
also like I said in a different comment, "pausing" in the context of a deployment does not freeze scaling to avoid fighting with hpa and is only meant to freeze upgrades so "pausing" for a replicaset is a totally new concept.

xiang90 replied:
> sounds like a different proposal?

yes. if we want to consider this approach we need to have another proposal as mentioned in appraoch1 vs approach2 section below.

xiang90 replied:
> also like I said in a different comment, "pausing" in the context of a deployment does not freeze scaling to avoid fighting with hpa and is only meant to freeze upgrades so "pausing" for a replicaset is a totally new concept.

it is a different concept. one is for pausing the control entirely, the other is for pausing a specific controlling action.

jian-he commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+Today, Deployment controller, while doing rolling update, will teardown all the pods and then recreate all the pods.
+This approach works in most cases but has several limitations:
+* The re-created pod may be allocated onto different nodes, but certain pods depend on local state on the node.
+* The new pod may not be allocated due to resources taken by other pods.
```
right, IMHO,  in-place update (without re-schedule a new pod) can avoid such additional scheduling loop.

jian-he commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+This approach works in most cases but has several limitations:
+* The re-created pod may be allocated onto different nodes, but certain pods depend on local state on the node.
+* The new pod may not be allocated due to resources taken by other pods.
+* The IP address may be changed due to the pod is re-created, but we want fixed IP across container upgrades.
```
Yes, that's why we want to avoid re-creating pods, to avoid changing the ip

janetkuo replied:
If the node is gone, the pods on that node will be gone. You cannot guarantee a fixed Pod IP.

jian-he commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+    // Reduce oldReplicaSet by the BatchSize
+    update(oldReplicaSet, replicas - BatchSize)
+    
+    // update pod controllerRef to the new ReplicaSet
```
Good point,  I missed this in the implementation, this will require additional changes to decouple it.
How about approach 2 ？

jian-he replied:
Actually, I rechecked the code (see below). IIUC,  if the existing pod already has a controllerRef, it compares its controllerRef with current controller's UID. In this case, the  existing pod's controllerRef != newReplicaSet UID, it returns false. Hence, in this scenario, pod's controllerRef will not be automatically updated to newReplicaSet.

```
	if controllerRef != nil {
		if controllerRef.UID != m.Controller.GetUID() {
			// Owned by someone else. Ignore.
			return false, nil
		}
```

jian-he commented on keps/sig-apps/0010-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,157 @@
+## Motivation
+Today, Deployment controller, while doing rolling update, will teardown all the pods and then recreate all the pods.
+This approach works in most cases but has several limitations:
+* The re-created pod may be allocated onto different nodes, but certain pods depend on local state on the node.
```
I will gather those and update the document.

helinbo2015 replied:
@janetkuo @jian-he I can share a use case of me. Previously i have engaged in a containerization project for wireless station applications, the application eats more memory (64GB), and there is no extra memory in the upgrade environment to initiate instances that may increase during a rolling upgrade.

jian-he commented:
/cc @janetkuo @kargakis I have updated the design doc. Please take a look, thanks. 
I moved the approach 1 into alternatives because based on discussion, the amount of changes is more significant, and we can focus on approach 2, and see if that makes sense. Thanks !

kargakis commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,162 @@
+
+// Spec to control the desired behavior of in-place update
++ type InPlaceUpdateDeployment struct {
++	 // The number of pods to be updated in each batch.
```
Make explicit in the doc that this is gated by readiness.

jian-he replied:
updated.

kargakis commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,162 @@
++	 // The number of pods to be updated in each batch.
++	 // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
++	 // Absolute number is calculated from percentage by rounding down.
++	 BatchSize intstr.IntOrString
```
How about using `MaxUnavailable` since this is a known concept in deployments already?

jian-he replied:
sounds good, I'll rename it to MaxUnavailable

kargakis commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,162 @@
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+
+    // swap oldReplicaSet and newReplicaSet information such as spec, revision, annotation etc.
```
What happens at this point needs to be spelled out more explicitly.

1. `copyInfo = oldReplicaSet`

What does this actually mean? Are you creating a new replica set in etcd that is essentially a copy of the old that you want to save? If so, what happens when you cannot create the replica set because of quota?

2. `oldReplicaSet = newReplicaSet`

When the old replica set is updated to the new replica set, at that point you have two replica sets that match the deployment spec. You may need to ensure that [`FindNewReplicaSet`](https://github.com/kubernetes/kubernetes/blob/440b2b6ea77e01fb4e2ad99c0b25ac7993e0e00f/pkg/controller/deployment/util/deployment_util.go#L650) handles that case correctly when the new strategy is used.

3. `newReplicaSet = copyInfo`

Assuming you have created the extra copy, are you going to clean it up or continue using it in subsequent rollouts?

kargakis replied:
How about storing the copy as an annotation in `newReplicaSet`?

jian-he replied:
1. ```What does this actually mean? Are you creating a new replica set in etcd that is essentially a copy of the old that you want to save? If so, what happens when you cannot create the replica set because of quota?```
Your proposal to store the copy as an annotation makes sense.
2. ```When the old replica set is updated to the new replica set, at that point you have two replica sets that match the deployment spec. You may need to ensure that FindNewReplicaSet handles that case correctly when the new strategy is used.```
Good point.   Since functions like "FindNewReplicaSet" sorts ReplicaSet by timestamp and assume current ReplicaSet has the latest timestamp, we may also need to swap the timestamp of the oldReplicaSet and new ReplicaSet

3. ```Assuming you have created the extra copy, are you going to clean it up or continue using it in subsequent rollouts?```
Keeping it for the purpose of rollback, the copy basically serves as a historic info for rollback.

kargakis commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
````diff
@@ -0,0 +1,162 @@
+
+```
+This approach requires some changes in ReplicaSet controller to support pause semantics. Another way is to pass certain information into ReplicaSet and prevent it from prematurely killing
+the pods while updating the pods' controllerRef and ReplicaSet's replicas. Or make Kubernetes support transactions which is another big effort.
````
Not sure how transactions would help. Mind expanding all alternatives into separate sections and elaborating as much as you can on each of them?

jian-he replied:
Yes, add in the doc

jian-he commented:
@kargakis thanks for your review, updated the doc again. could you give another look ? thanks !

kargakis commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,198 @@
+    // retrieve back the annotation for oldReplicaSet and replace newReplicaSet with the oldReplicaSet info.
+    // The oldReplicaSet info serves as a historic ReplicaSet for the purpose of rollback.
+    newReplicaSet = getAnnotation(oldReplicaSet)
+        
```
You also need to add the last step of doing the in-place update of the pods

jian-he replied:
Yep.  I had that in the beginning. Now I moved that to the end, as I think we should update the ReplicaSet info first, in case there's pod failed, a new pod created should be a pod with the new spec instead of old.

smarterclayton commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,198 @@
+## Proposal
+
+### API Change
+Augment deployment controller to have a new update strategy type: InPlaceUpdate.
```
Nit: UpdatePods would probably be more accurate

jian-he replied:
Do you suggest rename "InPlaceUpdate" to "UpdatePods" ? I was trying to follow the existing naming pattern "RollingUpdate".  or how about "InPlacePodUpdate" ?  I'm ok either way though.

smarterclayton commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,198 @@
+
+#### Update
+Below is the proposed approach for in-place update.
+Wait until expected number of pods created, when deployment controller receives the in-place update request
```
You need to describe what happens when the update does not work in place - ie a user changes resources or adds a container.

jian-he replied:
Good point. I think  we should throw error at the API level that changing resources while updateStrategyType is InPlaceUpdate is not allowed.  Similarly for adding a container.

jian-he commented:
@smarterclayton , @kargakis , thanks for your review ! I updated the doc again.  I havn't changed the "InPlaceUpdate" naming yet. I'll update it once we are on agreement.

bgrant0607 commented:
Ref https://github.com/kubernetes/kubernetes/issues/9043

bgrant0607 commented:
Most properties of pods aren't mutable. If this requires changing that, the node team must sign off.

Also, it's a complex feature, so reviewers need to agree in advance that they can make time.

And it would a big change for users, so we'd need to think about user impact.

bgrant0607 commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,204 @@
+    the network environment is not consistent, it becomes hard to guarantee a stable result.
+* The new pod may not be allocated due to resources taken by other pods. Preemption can mitigate this situation, but this 
+introduces an extra scheduling loop that can be avoided.
+* The IP address may be changed due to the pod is re-created, but we want fixed IP across container upgrades. 
```
I don't think this is a reasonable property to ensure in the Deployment controller

bgrant0607 replied:
This probably would make more sense in StatefulSet.

bgrant0607 commented:
Given the motivations for this, SIG Network and SIG Storage would need to look at this also

bgrant0607 commented:
I'm also not sure that CRI supports the kinds of updates necessary, other than image updates.

jian-he commented:
@bgrant0607 , thanks for your comments ! 
 
> Also, it's a complex feature, so reviewers need to agree in advance that they can make time.

Yep, really appreciate all the reviewers give so many good comments so far ! 

> And it would a big change for users, so we'd need to think about user impact.

This will be an opt-in feature. Only take effect if user explicitly specify  the inPlaceUpdate DeploymentStrategy.  If not set, nothing should be impacted.

> Most properties of pods aren't mutable. If this requires changing that, the node team must sign off.

In fact, current proposal does not change any fields of the pods. But to update the ReplicaSet. 

> I don't think this is a reasonable property to ensure in the Deployment controller

The main motivation for us is to support upgrade without reshuffling the pods. Reshuffling the pods can cause many issues at Alibaba scale. I can add more use case.
IMHO, since deployment controller controls the rollingUpdate process and , hence the proposal to augment deployment controller to support one more strategy.

> I'm also not sure that CRI supports the kinds of updates necessary, other than image updates.
We are indeed focusing on image update at this point. I can make this clear in the doc.

adohe commented:
```
I'm also not sure that CRI supports the kinds of updates necessary, other than image updates.
```
Just update image should not be problem for CRI, it is accomplished through `stop-create-start`.

bgrant0607 commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,215 @@
+
+
+## Motivation
+Today, Deployment controller, while doing rolling update, will teardown all the pods and then recreate all the pods. In essence, it reshuffles all the pods.
```
Do you have maxSurge set to zero?

bgrant0607 commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,215 @@
+This approach works in most cases but has several limitations:
+* we have thousands of pods running super critical user-facing online services with high availability guarantee. Re-shuffling all the pods over thousands of machines to do the upgrade is problematic:
+    1) Since pods are reshuffled, network topology for all the pods become indeterministic every time we do upgrade and this can cause inconsistent network performance.
+    It may also cause inconsistent runtime performance due to host hardware are different. This breaks user experience.
```
If you have heterogeneous hardware, this seems like a scheduling problem, not a controller issue.

bgrant0607 commented on keps/sig-apps/0012-20180522-deployment-inplace-update.md:
```diff
@@ -0,0 +1,215 @@
+* we have thousands of pods running super critical user-facing online services with high availability guarantee. Re-shuffling all the pods over thousands of machines to do the upgrade is problematic:
+    1) Since pods are reshuffled, network topology for all the pods become indeterministic every time we do upgrade and this can cause inconsistent network performance.
+    It may also cause inconsistent runtime performance due to host hardware are different. This breaks user experience.
+    2) Certain container images are very large, re-shuffling all the pods means re-downloading the images onto the hosts if the image is not present. 
```
The scheduler attempts to favor placement on nodes where the image is already present. Have you analyzed why that is not effective?

bgrant0607 commented:
@jian-he 

I'm sorry, but I still don't think the motivations cited by the proposal make sense. It sounds like other root causes are at work. I also think they are contrary to the way Deployment is intended to work. I'd be more in favor of increasing stability in StatefulSet before Deployment.

But we have much more basic usability issues that affect many, many more users that need to be addressed before this. And we need to improve test coverage. And address technical debt.

I don't think we can tackle this now. It would require significant bandwidth from the workload API maintainers. 

And I see that you haven't done significant work in kubernetes/kubernetes.

I suggest that you get to learn the system better for at least a couple quarters by fixing bugs and writing tests before proposing a major, complex feature.
https://github.com/kubernetes/kubernetes/pulls/jian-he

cc @kow3ns

jian-he commented:
@bgrant0607 
> I'd be more in favor of increasing stability in StatefulSet before Deployment.

I'm happy to contribute to Kubernetes in general. Do you think this is a useful feature for StatefulSet ? we are willing to contribute that too.

bgrant0607 commented:
@jian-he Thanks. If you would like to help, more/better tests would be much more valuable to the project right now than implementing new features.
